### ISSUE TITLE
kubens not resolving file paths relative to config file

### ISSUE THREAD

**Issue by Skymirrh** [ahmetb/kubectx#229]
Hi,

Just switched from bash version to Go version on a Windows environment (Git Bash).

It looks like `kubens` does not treat relative file paths in `.kube/config` as relative to the config file, but rather as relative to current working directory:
```
λ kubens
error: could not list namespaces (is the cluster accessible?): failed to initialize config: invalid configuration: [
unable to read client-cert REDACTED_PATH_CRT.crt for REDACTED_USER due to open REDACTED_PATH_CRT.crt: The system cannot find the file specified.,
unable to read client-key REDACTED_PATH_KEY.key for REDACTED_USER due to open REDACTED_PATH_KEY.key: The system cannot find the file specified.,
unable to read certificate-authority REDACTED_PATH_CRT_2.crt for REDACTED_CLUSTER due to open REDACTED_PATH_CRT_2.crt: The system cannot find the file specified.]

λ cd ~/.kube/
λ kubens
{works as expected}

λ cd ..
error: could not list namespaces (is the cluster accessible?): failed to initialize config: invalid configuration: [
unable to read client-cert REDACTED_PATH_CRT.crt for REDACTED_USER due to open REDACTED_PATH_CRT.crt: The system cannot find the file specified.,
unable to read client-key REDACTED_PATH_KEY.key for REDACTED_USER due to open REDACTED_PATH_KEY.key: The system cannot find the file specified.,
unable to read certificate-authority REDACTED_PATH_CRT_2.crt for REDACTED_CLUSTER due to open REDACTED_PATH_CRT_2.crt: The system cannot find the file specified.]
```
Current workaround is to use absolute file paths in config file by prefixing all paths with `C:\Users\USERNAME\.kube\`, but that kinda defeats the purpose of having a `.kube` directory :D

I did not have a chance to play with Go yet, but if nobody picks this up and I have some time this weekend, I'll have a try at it.

Thanks for the awesome work by the way! 👍

**Comment by Skymirrh** [comment]
Preliminary investigation indicates the errors occurs when calling `clientcmd` here:
https://github.com/ahmetb/kubectx/blob/401188fefd4fa3dcd94ee6a8eb1bb1efbff04b25/cmd/kubens/list.go#L61

So perhaps this actually is a bug for https://github.com/kubernetes/client-go/?

**Comment by ahmetb** [comment]
Does the same kubeconfig file work with kubectl?

My guess also would be that this is a client-go issue. That said I have never seen file paths in kubeconfig file. So I am not sure what you’re trying to do is valid.

**Comment by Skymirrh** [comment]
Yes, works fine with `kubectl` no matter the current working dir. Here are examples from the documentation where you can see how a config can include client certificates and/or CA certificates: https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/

**Comment by ahmetb** [comment]
Yeah I am suspecting client-go must have a knob for this. But I have not seen it.

**Comment by eshepelyuk** [comment]
Opened an duplicate since experiencing the same issue, but question is why kubectl works in this case ?  Does it use other GO libraries ?

**Comment by ahmetb** [comment]
This was reported by couple more folks using mingw/msys2 systems (similar to git bash) in #233.  I'm now trying to figure out if this reproduces outside these shells (natively on windows).

@Skymirrh @eshepelyuk Is it possible you move these kubeconfig file + cert files to %USERPROFILE%\.kube\ and run `kubens.exe` directly from cmd.exe or powershell?

> why kubectl works in this case

I'm guessing because kubectl determines which kubeconfig files to parse itself, so it can save where it has read the file from. in our case, we do custom yaml parsing to prevent the node structure in the document, and we need to investigate if `client-go` has a way of specifying "relative directory" of where the file was read from separately.

I personally don't have any files referenced from my kubeconfig file so I don’t run into this, but I can see many others do have this.

**Comment by eshepelyuk** [comment]
@ahmetb i will try to move an play from within cmd.exe

for now - there's a simple workaround - just to fix manually ~/.kube/confif file and put absolute path there. 

So, the issue itself may be closed - if there be some docs explaning how to workaround.

**Comment by ahmetb** [comment]
> just to fix manually ~/.kube/config file and put absolute path there.

May I ask everyone in this issue: Where do these manifests with relative file paths come from? Is it from a particular cloud-provider, or is it hand-crafted? (The result won't change anything, I’m just curious.)

**Comment by eshepelyuk** [comment]
In my case - those scripts are provided by our infrastructure engineers.

**Comment by Skymirrh** [comment]
> Is it possible you move these kubeconfig file + cert files to %USERPROFILE%.kube\ and run `kubens.exe` directly from cmd.exe or powershell?

I'm not sure I understand the question, but if the question is "what happens if you run `kubens` from `cmd` or `Powershell` rather than from `Git Bash`?", then there is no difference.

`Git Bash`'s `~` refers to Windows' `%USERPROFILE%`, so my `.kube` and its config/certificates are already stored in the right place. I can use `kubectl` from `Git Bash` or `Cmd`/`PowerShell` interchangeably, the only advantage is that on `Git Bash` I can source completion scripts to alleviate resource discovery 😁

> in our case, we do custom yaml parsing to prevent the node structure in the document, so we pass client-go library a "parsed YAML"

Ow, I missed that during preliminary investigation. My bad, would have obviously been the primary suspect had I seen it!

> Where do these manifests with relative file paths come from? Is it from a particular cloud-provider, or is it hand-crafted? (The result won't change anything, I’m just curious.)

From the [Kubernetes documentation itself](https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/), as already linked above, which includes calls to `kubectl config` using relative paths.

You may note this interesting bit from the documentation:
> File and path references in a kubeconfig file are relative to the location of the kubeconfig file. File references on the command line are relative to the current working directory. In `$HOME/.kube/config`, relative paths are stored relatively, and absolute paths are stored absolutely.
> -- [Source](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/#file-references)

**Comment by ahmetb** [comment]
Yeah the last bit’s certainly valuable. Now we need to go investigate how we can make kubectl use a particular base path when we use the [`RESTConfigFromKubeConfig`](https://github.com/ahmetb/kubectx/blob/401188fefd4fa3dcd94ee6a8eb1bb1efbff04b25/cmd/kubens/list.go#L61) method with an in-memory byte array of the kubeconfig file contents.

**Comment by Skymirrh** [comment]
I have taken a quick look at `clientcmd` and it looks like [BuildConfigFromFlags](https://github.com/kubernetes/client-go/blob/master/tools/clientcmd/client_config.go#L553) is a good candidate for loading config files.

There actually is an example provided by Go client:
```go
var kubeconfig *string
if home := homeDir(); home != "" {
	kubeconfig = flag.String("kubeconfig", filepath.Join(home, ".kube", "config"), "(optional) absolute path to the kubeconfig file")
} else {
	kubeconfig = flag.String("kubeconfig", "", "absolute path to the kubeconfig file")
}
flag.Parse()

// use the current context in kubeconfig
config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
if err != nil {
	panic(err.Error())
}
```
Source: https://github.com/kubernetes/client-go/blob/master/examples/out-of-cluster-client-configuration/main.go

**Comment by ahmetb** [comment]
> [BuildConfigFromFlags](https://github.com/kubernetes/client-go/blob/master/tools/clientcmd/client_config.go?rgh-link-date=2020-05-28T19%3A46%3A58Z#L553) is a good candidate for loading config files.

It is ––except it double-reads the file. We already read it in-memory once for detailed YAML doc tree parsing, if we use that it’ll parse the file yet again.

Furthermore we need to be able to work with multiple kubeconfig files, and it's likely we'll implement some of that logic ourselves. I’ll take a closer look later on. Thanks for flagging.

**Comment by Skymirrh** [comment]
> It is ––except it double-reads the file. We already read it in-memory once for detailed YAML doc tree parsing, if we use that it’ll parse the file yet again. (Since we open the file for RW, it might cause issues e.g. on windows.)

Note: I am not an expert at all with `clientcmd`, however from my quick looks at the documentation and code, I don't think `clientcmd` is intended to be used for loading the config file from memory. I think it would probably be more fitting to the intended use case if, instead of reading the config file to memory to get a YAML tree, you'd load the config file using `clientcmd` and *then* use that as the basis for the rest of your program (RW access is supported by `clientcmd` configs).

> Furthermore we need to be able to work with multiple kubeconfig files, and it's likely we'll implement some of that logic ourselves. I’ll take a closer look later on. Thanks for flagging.

FYI, it seems `clientcmd` can handle that as well, using [`EnvVar from PathOptions`](https://github.com/kubernetes/client-go/blob/f099a72e140ab46a70ceb9337c75802c3a2be045/tools/clientcmd/config.go#L47) which are deferred to [`ConfigAccess`](https://github.com/kubernetes/client-go/blob/f099a72e140ab46a70ceb9337c75802c3a2be045/tools/clientcmd/config.go#L34) via [`GetLoadingPrecedence()`](https://github.com/kubernetes/client-go/blob/f099a72e140ab46a70ceb9337c75802c3a2be045/tools/clientcmd/config.go#L76) and then used in [`loaders` for the `Load()` method](https://github.com/kubernetes/client-go/blob/f099a72e140ab46a70ceb9337c75802c3a2be045/tools/clientcmd/loader.go#L71).

**Comment by reitzig** [comment]
> May I ask everyone in this issue: Where do these manifests with relative file paths come from? Is it from a particular cloud-provider, or is it hand-crafted? (The result won't change anything, I’m just curious.)

FWIW, I use [direnv](https://direnv.net/) to cleanly separate k8s configs between projects:

`.envrc`
```
export KUBECONFIG="/home/raphael/.kube/config:kube_config.yaml"
```

Here, using a relative path for the project config seems prudent: it's automatically correctly scoped and will not break when moving the project folder around. Continuing from there, storing certificates in the project folder, for instance, seems a natural choice, so `kube_config.yaml` will have relative paths to those as well.
